core/capabilities/triggers/logevent/store.go

```diff
@@ -0,0 +1,78 @@
+package logevent
+
+import (
+	"fmt"
+	"sync"
+)
+
+type RegisterCapabilityFn[T any, Resp any] func() (*T, chan Resp, error)
+
+// Interface of the capabilities store
+type CapabilitiesStore[T any, Resp any] interface {
+	Read(capabilityID string) (value *T, ok bool)
+	ReadAll() (values []*T)
+	Write(capabilityID string, value *T)
+	InsertIfNotExists(capabilityID string, fn RegisterCapabilityFn[T, Resp]) (chan Resp, error)
+	Delete(capabilityID string)
+}
+
+// Implementation for the CapabilitiesStore interface
+type capabilitiesStore[T any, Resp any] struct {
+	mu           sync.RWMutex
+	capabilities map[string]*T
+}
+
+var _ CapabilitiesStore[string, string] = (CapabilitiesStore[string, string])(nil)
+
+// Constructor for capabilitiesStore struct implementing CapabilitiesStore interface
+func NewCapabilitiesStore[T any, Resp any]() CapabilitiesStore[T, Resp] {
+	return &capabilitiesStore[T, Resp]{
+		capabilities: map[string]*T{},
+	}
+}
+
+func (cs *capabilitiesStore[T, Resp]) Read(capabilityID string) (value *T, ok bool) {
+	cs.mu.RLock()
+	defer cs.mu.RUnlock()
+	trigger, ok := cs.capabilities[capabilityID]
+	return trigger, ok
+}
+
+func (cs *capabilitiesStore[T, Resp]) ReadAll() (values []*T) {
+	cs.mu.RLock()
+	defer cs.mu.RUnlock()
+	vals := make([]*T, 0)
+	for _, v := range cs.capabilities {
+		vals = append(vals, v)
+	}
+	return vals
+}
+
+func (cs *capabilitiesStore[T, Resp]) Write(capabilityID string, value *T) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	cs.capabilities[capabilityID] = value
+}
+
+func (cs *capabilitiesStore[T, Resp]) InsertIfNotExists(capabilityID string, fn RegisterCapabilityFn[T, Resp]) (chan Resp, error) {
+	cs.mu.RLock()
+	_, ok := cs.capabilities[capabilityID]
+	cs.mu.RUnlock()
+	if ok {
+		return nil, fmt.Errorf("capabilityID %v already exists", capabilityID)
+	}
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	value, respCh, err := fn()
+	if err != nil {
+		return nil, fmt.Errorf("error registering capability: %v", err)
+	}
+	cs.capabilities[capabilityID] = value
```

Comment:
This can produce duplicate, overwritten entries. The redundant check from my comment inside of the write lock was intentional. It is necessary for correctness.

Reply:
Sure

Reply:
Did you push? This still looks the same

Reply:
Yeah, I've pushed the faster exit above. Is that what you meant?

Reply:
The fast path is fine, but the slow path must still check for existence again before inserting because another goroutine may have gotten the lock first.
https://github.com/smartcontractkit/chainlink/pull/14308#discussion_r1781011835

Reply:
Sure, added this check in slow path as well